### ISSUE TITLE
High CPU usage and lag during cursor positioning

### ISSUE THREAD

**Issue by fxha** [marktext/marktext#1648]
### Description

**NOTE:** This is a release blocking issue for v0.16.x.

Mark Text becomes laggy when setting the cursor to another paragraph in larger documents.

### Steps to reproduce

1. Build latest develop version and load [MARKDOWN_SYNTAX.md](https://github.com/marktext/marktext/blob/develop/docs/MARKDOWN_SYNTAX.md).
2.  The first lag should occur after the file was loaded.
2. (Scroll a bit and) set cursor in some paragraph - short freeze/lag should occur.
3. Set cursor to another paragraph - short freeze/lag should occur.
4. Move the cursor in the same paragraph - no lag occurs.

**Expected behavior:**

No freezes/lags should occur.

**Actual behavior:**

Setting a cursor in another paragraph freezes the editor for a short time in larger documents. The behavior may depend on computer performance.

**Link to an example: [optional]**

12% CPU usage := 100% on one CPU core. In the both examples, I set cursor and directly type afterwards.

v0.16.0-testing:

![mt_lag](https://user-images.githubusercontent.com/22716132/68413360-633ac080-018e-11ea-830e-2c16803e2727.gif)

v0.15.1:

![mt_lag_v0 15 1](https://user-images.githubusercontent.com/22716132/68413363-66ce4780-018e-11ea-9ffd-d3bd8fe17421.gif)

### Versions

- Mark Text: develop (dfc7390) due to numbered code block lines - that are currently disabled!
- Operating system: all

**Comment by fxha** [comment]
Numbered code block lines are temporary disabled due to this issue.

**Stack Trance:**

![mt_numered_code_block_overview](https://user-images.githubusercontent.com/22716132/68491071-73b07100-024a-11ea-9693-e5a0f5bc566f.png)

![mt_numbered_code_blocks](https://user-images.githubusercontent.com/22716132/68491070-73b07100-024a-11ea-8a68-d8c7831a969f.png)

**Comment by Jocs** [comment]
@fxha The Stack Trance above is at `code line number` enabled?

**Comment by fxha** [comment]
> The Stack Trance above is at code line number enabled?

Yes, that's right. There are to much `Recalculate Style` and tons of forced `Layout` operations. BTW `setCursor` is also a performance bottleneck because it consume twice as much time as the whole document rendering.

**Comment by Jocs** [comment]
>because it consume twice

What do you mean twice here?

**Comment by fxha** [comment]
@Jocs I think I found another issue: we render a document twice when loading the first file using editor event bus and muya force render the whole document. This issue doesn't occur when closing the tab and open the file again in the same instance.

![mt_inital_rendering](https://user-images.githubusercontent.com/22716132/68493743-50d38c00-024d-11ea-8208-03d8a3f4c8d8.png)

>  > it consume twice
> 
> What do you mean twice here?

`render` takes ~180ms and `setCursor` ~420ms (Chromiums `Recalculate Style` take ~50s and `Layout` the remaining).

**Comment by Jocs** [comment]
@fxha I move this issue to project 0.17.0, and will not release the line number feature in v0.16.0.

**Comment by Jocs** [comment]
Thank you for this detailed performance report. The editor engine has been completely rewritten since this was filed. In the new engine (@muyajs/core, now shipping): clicking between paragraphs no longer triggers a content re-render — the click handler defers via requestAnimationFrame and passes needUpdate=false to setCursor, so only the DOM Selection is updated. Code block line numbers are now reconciled O(delta) via ResizeObserver (which fires post-layout asynchronously), not on every cursor change. There is no longer a document-wide style recalculation on inter-paragraph cursor positioning. Closing as resolved by the engine migration.
